### ISSUE TITLE
coder server --wg-tunnel-host causes tunneld to return 404

### ISSUE THREAD

**Issue by hh** [coder/coder#9480]
TLDR: tunneld and coder server most recent versions don't seem to work together (OR I'm missing something obvious to others in the configuration of them)

*code server* attempt to use try.sharing.io as our wgtunnel host (with no local db or other settings):

```
rm -rf ~/Library/Application\ Support/coderv2/{url,devtunnel,postgres/data} && ./coder server --wg-tunnel-host try.sharing.io
```

Results in api/v2/clients 404

```
Coder v0.25.0+d843735 - Your Self-Hosted Remote Development Platform
Started HTTP listener at http://127.0.0.1:3000
Using built-in PostgreSQL (/Users/hh/Library/Application Support/coderv2/postgres)
Opening tunnel so workspaces can connect to your deployment. For production scenarios, specify an external access URL
Using tunnel in Unknown with latency  51.205ms .
Stopping built-in PostgreSQL...
Stopped built-in PostgreSQL
running command "coder server": create tunnel:                                          
    github.com/coder/coder/cli.(*RootCmd).Server.func1                                  
        /home/runner/actions-runner/_work/coder/coder/cli/server.go:339                 
  - initial client registration:                                                        
    github.com/coder/wgtunnel/tunnelsdk.(*Client).LaunchTunnel                          
        /home/runner/go/pkg/mod/github.com/coder/wgtunnel@v0.1.5/tunnelsdk/tunnel.go:191
  - POST https://try.sharing.io/api/v2/clients: unexpected status code 404: Not found.  
 ```

We take a look at the ingress-nginx logs and see the 404:

```
kubectl -n ingress-nginx logs svc/ingress-ngnix-ingress-nginx-controller --since=1m -f | grep tunneld
```

```
127.0.0.1 - - [01/Sep/2023:10:48:51 +0000] "POST /api/v2/clients HTTP/2.0" 404 25 "-" "Go-http-client/2.0" 208 0.000 [hh-tunneld-api-http] [] 10.0.0.145:12345 25 0.000 404 43fb8c7e1a64d271dd66d9f06890e36c
```

tunnelt is run from latest source with the full kubernetes config below:
```
./wgtunnel/build/tunneld  -v
```

```
2023-09-01 10:42:18.123 [INFO]  <main.go:247>   parsed private key      {"hash": "xxxxxxxxxxx"}
2023-09-01 10:42:18.125 [INFO]  <main.go:298>   listening for requests  {"listen_address": "0.0.0.0:12345"}
2023-09-01 10:42:18.125 [INFO]  <main.go:291>   starting pprof server   {"listen_address": "127.0.0.1:6060"}
``` 
 
Here is the wider config:
 
- [src](#org412e222)
  - [build](#org49c8c3f)
- [env settings](#orgaae7603)
  - [&#x2013;verbose, -v](#org9eb5e1d)
  - [&#x2013;listen-address value, -a value](#org85e03f6)
  - [&#x2013;base-url value, -u value](#org2ef0aa0)
  - [&#x2013;wireguard-endpoint value, &#x2013;wg-endpoint value](#org1bfdafe)
  - [&#x2013;wireguard-port value, &#x2013;wg-port value](#orge1523ab)
  - [&#x2013;wireguard-key wg genkey, &#x2013;wg-key wg genkey](#org17fb622)
  - [&#x2013;wireguard-key-file value, &#x2013;wg-key-file value](#org726e1a8)
  - [&#x2013;wireguard-mtu value, &#x2013;wg-mtu value](#org683392c)
  - [&#x2013;wireguard-server-ip value, &#x2013;wg-server-ip value](#org208fe30)
  - [&#x2013;wireguard-network-prefix value, &#x2013;wg-network-prefix value](#org155d5c1)
  - [&#x2013;real-ip-header value](#org28ff618)
  - [&#x2013;pprof-listen-address value](#org74aeee4)
  - [&#x2013;tracing-honeycomb-team value](#org48100f5)
  - [&#x2013;tracing-instance-id value](#org5a32628)
- [run](#orgc0dcdfe)
- [verify (without kubernetes)](#org81a91bf)
  - [listening on ip sockets](#org9c91351)
  - [api](#org2dbd2cd)
- [verify with k8s svc+cert+ingress](#org4b23387)
  - [top level](#orgc2b399c)
  - [api](#org93d2373)
- [tunneld wireguard service](#org38f36a3)
  - [tunneld-wg-svc.yaml](#org0a3f012)
  - [kubectl apply](#org684762a)
  - [kubectl describe](#org22ebfca)
- [tunneld api service](#org05063b3)
  - [tunneld-api-svc.yaml](#orgfba64dd)
  - [kubectl apply](#orgb2806f2)
  - [kubectl describe](#orgdbf8757)
- [wildcard cert](#org1738619)
  - [certificate resource yaml](#org2fcbc81)
  - [kubectl apply cert](#org2f2e179)
  - [kubectl describe cert](#org0186c2f)
  - [kubectl describe secret](#org1034085)
- [ingress](#orgfcc2ddd)
  - [ingress resource yaml](#orgfd2f3f9)
  - [kubectl apply ingress](#orga7a7cbe)
  - [kubectl describe ingress](#orge39a160)
- [endpoints](#org81c6a2c)
  - [get](#org59562a7)

<a id="org412e222"></a>

# src

<a id="org49c8c3f"></a>

## build

```tmate
git clone https://github.com/coder/wgtunnel
cd wgtunnel
make build/tunneld
```

<a id="orgaae7603"></a>

# env settings

All options can be set via ENV vars, so we do so. See **tunneld -h**

<a id="org9eb5e1d"></a>

## &#x2013;verbose, -v

Enable verbose logging. (default: false)

```shell
export TUNNELD_VERBOSE=true
```

<a id="org85e03f6"></a>

## &#x2013;listen-address value, -a value

HTTP listen address for the API and tunnel traffic. (default: &ldquo;127.0.0.1:8080&rdquo;)

```shell
export TUNNELD_LISTEN_ADDRESS="0.0.0.0:12345"
```

<a id="org2ef0aa0"></a>

## &#x2013;base-url value, -u value

The base URL to use for the tunnel, including scheme. All tunnels will be subdomains of this hostname.

```shell
export TUNNELD_BASE_URL="https://try.sharing.io:443"
```

<a id="org1bfdafe"></a>

## &#x2013;wireguard-endpoint value, &#x2013;wg-endpoint value

The UDP address advertised to clients that they will connect to for wireguard connections. It should be in the form host:port.

```shell
export TUNNELD_WIREGUARD_ENDPOINT="try.sharing.io:54321"
```

<a id="orge1523ab"></a>

## &#x2013;wireguard-port value, &#x2013;wg-port value

The UDP port that the wireguard server will listen on. It should be the same as the port in wireguard-endpoint. (default: 0)

```shell
export TUNNELD_WIREGUARD_PORT=54321
```

<a id="org17fb622"></a>

## &#x2013;wireguard-key wg genkey, &#x2013;wg-key wg genkey

The private key for the wireguard server. It should be base64 encoded. You can generate a key with wg genkey. Mutually exclusive with wireguard-key-file.

```shell
export TUNNELD_WIREGUARD_KEY=`wg genkey`
```

<a id="org726e1a8"></a>

## &#x2013;wireguard-key-file value, &#x2013;wg-key-file value

The file path containing the private key for the wireguard server. The contents should be base64 encoded. If the file does not exist, a key will be generated for you and written to the file. Mutually exclusive with wireguard-key.

```shell
export TUNNELD_WIREGUARD_KEYFILE=
```

<a id="org683392c"></a>

## &#x2013;wireguard-mtu value, &#x2013;wg-mtu value

The MTU to use for the wireguard interface. (default: 1280)

```shell
export TUNNELD_WIREGUARD_MTU=1280
```

<a id="org208fe30"></a>

## &#x2013;wireguard-server-ip value, &#x2013;wg-server-ip value

The virtual IP address of this server in the wireguard network. Must be an IPv6 address contained within wireguard-network-prefix. (default: &ldquo;fcca::1&rdquo;)

```shell
export TUNNELD_WIREGUARD_SERVER_IP=fcca::1
```

<a id="org155d5c1"></a>

## &#x2013;wireguard-network-prefix value, &#x2013;wg-network-prefix value

The CIDR of the wireguard network. All client IPs will be generated within this network. Must be a IPv6 CIDR and have at least 64 bits available. (default: &ldquo;fcca::/16&rdquo;)

```shell
export TUNNELD_WIREGUARD_NETWORK_PREFIX=fcca::/16
```

<a id="org28ff618"></a>

## &#x2013;real-ip-header value

Use the given header as the real IP address rather than the remote socket address.

```shell
export TUNNELD_REAL_IP_HEADER: "X-Forwarded-For"
```

<a id="org74aeee4"></a>

## &#x2013;pprof-listen-address value

The address to listen on for pprof. If set to an empty string, pprof will not be enabled. (default: &ldquo;127.0.0.1:6060&rdquo;)

```shell
# export TUNNELD_PPROF_LISTEN_ADDRESS=0.0.0.0:6060
export TUNNELD_PPROF_LISTEN_ADDRESS=
```

<a id="org48100f5"></a>

## &#x2013;tracing-honeycomb-team value

The Honeycomb team ID to send tracing data to. If not specified, tracing will not be shipped anywhere.

```shell
export  TUNNELD_TRACING_HONEYCOMB_TEAM=
```

<a id="org5a32628"></a>

## &#x2013;tracing-instance-id value

The instance ID to annotate all traces with that uniquely identifies this deployment. [$TUNNELD<sub>TRACING</sub><sub>INSTANCE</sub><sub>ID</sub>]

```shell
export  TUNNELD_TRACING_INSTANCE_ID=
```

<a id="orgc0dcdfe"></a>

# run

```shell
./wgtunnel/build/tunneld
```

<a id="org81a91bf"></a>

# verify (without kubernetes)

<a id="org9c91351"></a>

## listening on ip sockets

```shell
lsof -i -n -P | grep tunneld
```

```
tunneld   3757815   ii    3u  IPv4  477030064      0t0  UDP *:54321
tunneld   3757815   ii    7u  IPv6  477030065      0t0  UDP *:54321
tunneld   3757815   ii   11u  IPv6  477020952      0t0  TCP *:12345 (LISTEN)
```

<a id="org2dbd2cd"></a>

## api

```shell
curl -X POST http://localhost:12345/api/v2/clients
```

```
{"message":"Not found."}
```

<a id="org4b23387"></a>

# verify with k8s svc+cert+ingress

<a id="orgc2b399c"></a>

## top level

```shell
curl https://try.sharing.io
```

```
https://coder.com
```

<a id="org93d2373"></a>

## api

```shell
curl -X POST https://try.sharing.io/api/v2/clients
```

```
{"message":"Request body must be valid JSON.","detail":"EOF"}
```

<a id="org38f36a3"></a>

# tunneld wireguard service

<a id="org0a3f012"></a>

## tunneld-wg-svc.yaml

```yaml
apiVersion: v1
kind: Service
metadata:
  name: tunneld-wg
  namespace: hh
spec:
  type: ClusterIP
  internalTrafficPolicy: Cluster
  ipFamilyPolicy: SingleStack
  ports:
  - port: 54321
    protocol: UDP
    name: wg
    targetPort: 54321
  selector:
    spaceapp: iipod
    spacename: london
```

<a id="org684762a"></a>

## kubectl apply

```shell
kubectl apply -f tunneld-wg-svc.yaml
```

```
service/tunneld-wg created
```

<a id="org22ebfca"></a>

## kubectl describe

```shell
kubectl describe svc tunneld-wg
```

```
Name:              tunneld-wg
Namespace:         hh
Labels:            <none>
Annotations:       <none>
Selector:          spaceapp=iipod,spacename=london
Type:              ClusterIP
IP Family Policy:  SingleStack
IP Families:       IPv4
IP:                10.109.106.31
IPs:               10.109.106.31
Port:              wg  54321/UDP
TargetPort:        54321/UDP
Endpoints:         10.0.0.145:54321
Session Affinity:  None
Events:            <none>
```

<a id="org05063b3"></a>

# tunneld api service

<a id="orgfba64dd"></a>

## tunneld-api-svc.yaml

```yaml
apiVersion: v1
kind: Service
metadata:
  name: tunneld-api
  namespace: hh
spec:
  type: ClusterIP
  internalTrafficPolicy: Cluster
  ipFamilyPolicy: SingleStack
  ports:
  - port: 80
    name: http
    protocol: TCP
    targetPort: 12345
  selector:
    spaceapp: iipod
    spacename: london
```

<a id="orgb2806f2"></a>

## kubectl apply

```shell
kubectl apply -f tunneld-api-svc.yaml
```

```
service/tunneld-api created
```

<a id="orgdbf8757"></a>

## kubectl describe

```shell
kubectl describe svc tunneld-api
```

```
Name:              tunneld-api
Namespace:         hh
Labels:            <none>
Annotations:       <none>
Selector:          spaceapp=iipod,spacename=london
Type:              ClusterIP
IP Family Policy:  SingleStack
IP Families:       IPv4
IP:                10.109.159.218
IPs:               10.109.159.218
Port:              http  80/TCP
TargetPort:        12345/TCP
Endpoints:         10.0.0.145:12345
Session Affinity:  None
Events:            <none>
```

<a id="org1738619"></a>

# wildcard cert

<a id="org2fcbc81"></a>

## certificate resource yaml

```yaml
apiVersion: cert-manager.io/v1
kind: Certificate
metadata:
  name: try
  namespace: hh
spec:
  issuerRef:
    name: letsencrypt-pdns
    group: cert-manager.io
    kind: ClusterIssuer
  commonName: "*.try.sharing.io"
  secretName: try-tls
  dnsNames:
    - "try.sharing.io"
    - "*.try.sharing.io"
```

<a id="org2f2e179"></a>

## kubectl apply cert

```shell
kubectl apply -f cert.yaml
```

```
certificate.cert-manager.io/try created
```

<a id="org0186c2f"></a>

## kubectl describe cert

```shell
kubectl describe certs try
```

```
Name:         try
Namespace:    hh
Labels:       <none>
Annotations:  <none>
API Version:  cert-manager.io/v1
Kind:         Certificate
Metadata:
  Creation Timestamp:  2023-09-01T08:14:31Z
  Generation:          2
  Resource Version:    1495446
  UID:                 339b44e6-51ae-4dfa-9521-2cbd67c3b801
Spec:
  Common Name:  *.try.sharing.io
  Dns Names:
    try.sharing.io
    *.try.sharing.io
  Issuer Ref:
    Group:      cert-manager.io
    Kind:       ClusterIssuer
    Name:       letsencrypt-pdns
  Secret Name:  try-tls
Status:
  Conditions:
    Last Transition Time:  2023-09-01T08:19:05Z
    Message:               Certificate is up to date and has not expired
    Observed Generation:   2
    Reason:                Ready
    Status:                True
    Type:                  Ready
  Not After:               2023-11-30T07:19:03Z
  Not Before:              2023-09-01T07:19:04Z
  Renewal Time:            2023-10-31T07:19:03Z
  Revision:                2
Events:
  Type    Reason     Age                  From                                       Message
  ----    ------     ----                 ----                                       -------
  Normal  Generated  5m6s                 cert-manager-certificates-key-manager      Stored new private key in temporary Secret resource "try-zx5df"
  Normal  Requested  5m6s                 cert-manager-certificates-request-manager  Created new CertificateRequest resource "try-c2tvg"
  Normal  Issuing    35s (x2 over 5m7s)   cert-manager-certificates-trigger          Issuing certificate as Secret does not exist
  Normal  Generated  35s                  cert-manager-certificates-key-manager      Stored new private key in temporary Secret resource "try-2dg7p"
  Normal  Requested  35s                  cert-manager-certificates-request-manager  Created new CertificateRequest resource "try-v7dkp"
  Normal  Issuing    33s (x2 over 2m49s)  cert-manager-certificates-issuing          The certificate has been successfully issued
```

<a id="org1034085"></a>

## kubectl describe secret

```shell
kubectl describe secret try-tls
```

```
Name:         try-tls
Namespace:    hh
Labels:       controller.cert-manager.io/fao=true
Annotations:  cert-manager.io/alt-names: *.try.sharing.io,try.sharing.io
              cert-manager.io/certificate-name: try
              cert-manager.io/common-name: *.try.sharing.io
              cert-manager.io/ip-sans:
              cert-manager.io/issuer-group: cert-manager.io
              cert-manager.io/issuer-kind: ClusterIssuer
              cert-manager.io/issuer-name: letsencrypt-pdns
              cert-manager.io/uri-sans:

Type:  kubernetes.io/tls

Data
====
tls.crt:  5542 bytes
tls.key:  1679 bytes
```

<a id="orgfcc2ddd"></a>

# ingress

<a id="orgfd2f3f9"></a>

## ingress resource yaml

```yaml
---
apiVersion: networking.k8s.io/v1
kind: Ingress
metadata:
  annotations:
    nginx.ingress.kubernetes.io/proxy-read-timeout: "3600"
    nginx.ingress.kubernetes.io/proxy-send-timeout: "3600"
    nginx.ingress.kubernetes.io/server-snippets: |
      location / {
      proxy_set_header Upgrade $http_upgrade;
      proxy_http_version 1.1;
      proxy_set_header X-Forwarded-Host $http_host;
      proxy_set_header X-Forwarded-Proto $scheme;
      proxy_set_header X-Forwarded-For $remote_addr;
      proxy_set_header Host $host;
      proxy_set_header Connection "upgrade";
      proxy_cache_bypass $http_upgrade;
      }
  name: try
  namespace: hh
spec:
  ingressClassName: nginx
  rules:
    - host: "try.sharing.io"
      http:
        paths:
          - backend:
              service:
                name: tunneld-api
                port:
                  name: http
            path: /
            pathType: Prefix
    - host: "*.try.sharing.io"
      http:
        paths:
          - backend:
              service:
                name: tunneld-api
                port:
                  name: http
            path: /
            pathType: Prefix
  tls:
    - hosts:
        - "try.sharing.io"
        - "*.try.sharing.io"
      secretName: try-tls
```

<a id="orga7a7cbe"></a>

## kubectl apply ingress

```shell
kubectl apply -f ingress.yaml
```

```
ingress.networking.k8s.io/try created
```

<a id="orge39a160"></a>

## kubectl describe ingress

```shell
kubectl describe ingress try
```

```
Name:             try
Labels:           <none>
Namespace:        hh
Address:          145.40.113.253
Ingress Class:    nginx
Default backend:  <default>
TLS:
  try-tls terminates try.sharing.io,*.try.sharing.io
Rules:
  Host              Path  Backends
  ----              ----  --------
  try.sharing.io
                    /   tunneld-api:http (10.0.0.145:12345)
  *.try.sharing.io
                    /   tunneld-api:http (10.0.0.145:12345)
Annotations:        nginx.ingress.kubernetes.io/proxy-read-timeout: 3600
                    nginx.ingress.kubernetes.io/proxy-send-timeout: 3600
                    nginx.ingress.kubernetes.io/server-snippets:
                      location / {
                      proxy_set_header Upgrade $http_upgrade;
                      proxy_http_version 1.1;
                      proxy_set_header X-Forwarded-Host $http_host;
                      proxy_set_header X-Forwarded-Proto $scheme;
                      proxy_set_header X-Forwarded-For $remote_addr;
                      proxy_set_header Host $host;
                      proxy_set_header Connection "upgrade";
                      proxy_cache_bypass $http_upgrade;
                      }
Events:
  Type    Reason  Age               From                      Message
  ----    ------  ----              ----                      -------
  Normal  Sync    3s (x4 over 30m)  nginx-ingress-controller  Scheduled for sync
```

<a id="org81c6a2c"></a>

# endpoints

<a id="org59562a7"></a>

## get

```shell
kubectl get endpoints | grep tunneld
```

```
tunneld-api                       10.0.0.145:12345                               49m
tunneld-wg                        10.0.0.145:54321                               47m
```

**Comment by hh** [comment]
Not sure what's up with my config, but I'll writeup a post on deploying tunneld in Kubernetes. I got it working.
